jerry-core/include/jerryscript-comp.h

```diff
@@ -0,0 +1,56 @@
+/* Copyright JS Foundation and other contributors, http://js.foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef JERRYSCRIPT_COMP_H
+#define JERRYSCRIPT_COMP_H
+
+#ifdef __cplusplus
+extern "C"
+{
+#endif /* __cplusplus */
+
+/** \addtogroup jerry-comp Jerry compiler compatibility component
+ * @{
+ */
+
+/*
+ * Attributes
+ */
+#define __attr_aligned___(ALIGNMENT) __attribute__((aligned(ALIGNMENT)))
+#define __attr_always_inline___ __attribute__((always_inline))
+#define __attr_const___ __attribute__((const))
+#define __attr_deprecated___ __attribute__((deprecated))
+#define __attr_format___(...) __attribute__((format(__VA_ARGS__)))
+#define __attr_hot___ __attribute__((hot))
+#define __attr_noinline___ __attribute__((noinline))
+#define __attr_noreturn___ __attribute__((noreturn))
+#define __attr_pure___ __attribute__((pure))
+#define __attr_section___(SECTION) __attribute__((section(SECTION)))
+#define __attr_warn_unused_result___ __attribute__((warn_unused_result))
+
+/*
+ * Condition likeliness, unlikeliness
+ */
+#define likely(x) __builtin_expect(!!(x), 1)
+#define unlikely(x) __builtin_expect(!!(x), 0)
```

Comment:
The original code has many ifdefs. Why do we remove them? __builtin_expect is very much gcc sepcific.

Reply:
Actually, not that many (or not that many useful). The attribute-related ifndefs are legacy to be removed, I'm quite sure. I'v tried to track back their origin last week and it turned out that they are leftovers from some 3.5+ years. They don't seem to be any useful but nobody dared to touch them, like ever. Now I dare...

As for the `__builtin_expect` issue. It is partially right that I have been sloppy, but it was also partially intentional. The builtins are compiler dependent indeed, but our current macro guards are imprecise / overly strict. E.g., clang also has that builtin, but the existing `__GNUC__` guard will exclude clang, if I'm not mistaken.

So, for now at least, I deliberately dropped the macro guards as gcc and clang are the officially maintained and tested compilers (Travis CI is green!), accepting that we may see some follow up patches from hopefully-not-too-upset users of different compilers. But this time, we will hopefully have more precise guards and only where they are needed and all in one place. At least, that was my motivation here.